**<span style="color:#448844">Note</span>** This notebook is meant to be interactive. Launch this notebook in Jupyter to see its full potential.

# Naive Bayes Exercise

In this notebook, you will learn to implement a Naive Bayes classifier using sklearn. We will be creating two classifiers, one which assumes a Gaussian distribution, and another that assumes a multinomial distrbution.

## Instructions for All Labs
* Read each cell and implement the TODOs sequentially. The markdown/text cells also contain instructions which you need to follow to get the whole notebook working.
* Do not change the variable names unless the instructor allows you to.
* Some markdown cells contain questions.
  * For questions <span style="color:red;">colored in red</span>, you must submit your answers in the corresponding Assignment in the course page. Make sure that you enter your responses in the item with the matching question code. Answers that do not follow the prescribed format will automatically be marked wrong by the checker.
  * For questions <span style="color:green;">colored in green</span>, you don't have to submit your answers, but you must think about these questions as they will help enrich your understanding of the concepts covered in the labs.
* You are expected to search how to some functions work on the Internet or via the docs. 
* You may add new cells for "scrap work".
* The notebooks will undergo a "Restart and Run All" command, so make sure that your code is working properly.
* You may not reproduce this notebook or share them to anyone.

In [1]:
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
plt.style.use('ggplot')

plt.rcParams['figure.figsize'] = (12.0, 8.0) # set default size of plots
plt.rcParams['image.interpolation'] = 'nearest'

# Fix the seed of the random number 
# generator so that your results will match ours
np.random.seed(1)

%load_ext autoreload
%autoreload 2

# Gaussian Naive Bayes


Our first dataset (iris dataset), our assumption is our data follows a Gaussian distribution.

**Dataset:**

Our first data set is the iris data set which contains 3 classes of 50 instances each. Each class refers to a type of iris plant.

Attribute Information:

1. Sepal length in cm
2. Sepal width in cm
3. Petal length in cm
4. Petal width in cm
5. Class (Species):
    - Iris Setosa
    - Iris Versicolour
    - Iris Virginica

## Preprocessing our data

Let's load the iris dataset.

In [2]:
import pandas as pd

iris = pd.read_csv('iris.csv')
iris.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [3]:
iris.groupby('species').describe()

sepal_length                                              \
                  count   mean       std  min    25%  50%  75%  max   
species                                                               
setosa             50.0  5.006  0.352490  4.3  4.800  5.0  5.2  5.8   
versicolor         50.0  5.936  0.516171  4.9  5.600  5.9  6.3  7.0   
virginica          50.0  6.588  0.635880  4.9  6.225  6.5  6.9  7.9   

           sepal_width         ... petal_length      petal_width         \
                 count   mean  ...          75%  max       count   mean   
species                        ...                                        
setosa            50.0  3.418  ...        1.575  1.9        50.0  0.244   
versicolor        50.0  2.770  ...        4.600  5.1        50.0  1.326   
virginica         50.0  2.974  ...        5.875  6.9        50.0  2.026   

                                               
                 std  min  25%  50%  75%  max  
species                                        
setosa      0.107210  0.1  0.2  0.2  0.3  0.6  
versicolor  0.197753  1.0  1.2  1.3  1.5  1.8  
virginica   0.274650  1.4  1.8  2.0  2.3  2.5  

[3 rows x 32 columns]

Right now, we have to convert our nominal labels (word labels: setosa, versicolor, viriginica) into numerical labels (number labels: 0,1,2; 0 for setosa, 1 for versicolor, 2 for viriginica)

In [4]:
from sklearn import preprocessing

This will convert the unique nominal values there are in iris["species"] to a unique number

In [5]:
label_enc = preprocessing.LabelEncoder()
label_enc.fit(iris["species"])

LabelEncoder()

We can check the original labels

This will transform the list to match the numerical code mapping

In [6]:
label_enc.transform(iris["species"])

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

Let's see the mapping of the original nominal labels and the numerical codes

In [7]:
print("Original labels:", label_enc.classes_, "\n")

print("Mapping from nominal to numerical labels:")
print(dict(zip(label_enc.classes_,label_enc.transform(label_enc.classes_))))

Original labels: ['setosa' 'versicolor' 'virginica'] 

Mapping from nominal to numerical labels:
{'setosa': np.int64(0), 'versicolor': np.int64(1), 'virginica': np.int64(2)}


Now that we have the numerical encoding and the mapping, we can now change the `species` column to its numerical mapping

In [8]:
iris["species"] = label_enc.transform(iris["species"])

Let's see the results now:

In [9]:
iris.head()

sepal_length  sepal_width  petal_length  petal_width  species
0           5.1          3.5           1.4          0.2        0
1           4.9          3.0           1.4          0.2        0
2           4.7          3.2           1.3          0.2        0
3           4.6          3.1           1.5          0.2        0
4           5.0          3.6           1.4          0.2        0

Like in the previous notebooks, we will separate our `X` from our target `y` (species). 


__Note__: `iris.values[:,:-1]` will get all rows, and all columns except for the last column


__Note__: `iris.values[:,-1]` will get the last column only. We set the the labels as integers because its default data type is float.

In [10]:
X = iris.values[:,:-1]
y = iris.values[:,-1].astype(int)

In [11]:
print("X shape: ", X.shape)
print("y shape: ", y.shape)

X shape:  (150, 4)
y shape:  (150,)


Let's separate the training from the test set. 

Set the test size to `0.3`, make sure to stratify based on `species`/`y`. Also set the `random_state` to `42` so our results match.

In [12]:
from sklearn.model_selection import train_test_split
# train_test_split?

In [13]:
# write code here
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, stratify=y, random_state=42)

In [14]:
X_train.shape

(105, 4)

**Sanity Check:** X_train should have a shape of `(105, 4)`

## Building our model
Because our features `X` are continuous values, we will use `sklearn`'s `GaussianNB` model.

In [15]:
from sklearn.naive_bayes import GaussianNB

Intialize a `GaussianNB` model

In [16]:
iris_nb = GaussianNB()

Train it

In [17]:
iris_nb.fit(X_train, y_train)

GaussianNB()

And, get its training predictions

In [18]:
predictions = iris_nb.predict(X_train)
predictions

array([1, 1, 0, 2, 1, 2, 0, 0, 0, 2, 2, 0, 0, 1, 1, 2, 0, 0, 2, 2, 0, 2,
       2, 2, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 2, 2, 0, 2, 0, 1, 0,
       2, 1, 0, 2, 1, 2, 1, 0, 1, 2, 1, 2, 0, 1, 0, 1, 1, 1, 2, 1, 1, 2,
       2, 0, 2, 1, 1, 2, 0, 2, 2, 1, 0, 2, 2, 0, 0, 2, 2, 2, 0, 2, 1, 2,
       2, 0, 1, 1, 1, 1, 1, 0, 2, 1, 2, 0, 0, 1, 0, 1, 0])

We will be computing for the accuracy multiple times in this notebook, so let's create a function for this.

`compute_accuracy()` will compute for the accuracy given two vectors of equal length

__Inputs:__
- `predictions`: A numpy array of shape `(N,)` consisting of `N` samples representing the predicted values
- `actual`: A numpy array of shape `(N,)` consisting of `N` samples representing the actual (target) values

__Outputs:__
- `accuracy`: A scalar representing the percentage of elements where `predictions` and `actual` match out of the total number of elements

In [19]:
def compute_accuracy(predictions, actual):
    predictions = np.array(predictions)
    actual = np.array(actual)
    accuracy = np.mean (predictions == actual)
    return accuracy

Let's see how well our model performed on the training data

In [20]:
print("Training accuracy: ", compute_accuracy(predictions, y_train), "%")

Training accuracy:  0.9809523809523809 %


That's a good result. Let's see if it will perform well on our test set.

In [21]:
predictions = iris_nb.predict(X_test)
predictions

array([2, 1, 1, 1, 2, 2, 1, 1, 0, 2, 0, 0, 2, 2, 0, 2, 1, 0, 0, 0, 1, 0,
       1, 2, 2, 1, 1, 1, 1, 0, 1, 2, 1, 0, 2, 0, 0, 0, 0, 2, 1, 0, 1, 2,
       1])

In [22]:
print("Test accuracy: ", compute_accuracy(predictions, y_test), "%")

Test accuracy:  0.9111111111111111 %


**Sanity Check:** You should get a 91.111% accuracy

## Checking the learned parameters
We can also peer into the parameters the model learned.

This is how you get the number of instances (of each class) the model received as the training set

In [23]:
iris_nb.class_count_

array([35., 35., 35.])

You can also get the priors the model learned

In [25]:
# write code here
iris_nb.class_prior_


array([0.33333333, 0.33333333, 0.33333333])

<span style="color:red;">**Question 7-1:** What is the prior computed for the first class? Round of to four decimal places.</span>

<span style="color:red;">**Question 7-2:** What is the prior computed for the second class? Round of to four decimal places.</span>

<span style="color:red;">**Question 7-3:** What is the prior computed for the second class? Round of to four decimal places.</span>

<span style="color:green;">**Question:** How are the priors calculated?</span>

Gaussian Naive Bayes classifiers have **`k * d * 2`** number of parameters (not including the priors)

> where <br>
> **`k`** - number of classes <br>
> **`d`** - number of dimensions/features <br>
> **`2`** - because we calculate for the means and variances of each feature <br>

Get the computed means of the model

In [28]:
# write code here
iris_nb.theta_

array([[4.98857143, 3.41142857, 1.48857143, 0.23714286],
       [5.94857143, 2.73142857, 4.23714286, 1.30857143],
       [6.68285714, 3.00857143, 5.63142857, 2.06857143]])

<span style="color:red;">**Question 7-4:** What is the shape of the computed means?</span>

Get the computed variances of the model

In [31]:
# write code here
iris_nb.var_

array([[0.10329796, 0.17586939, 0.02272653, 0.00976327],
       [0.24078368, 0.08558368, 0.21147755, 0.03564082],
       [0.42484898, 0.11735511, 0.32272653, 0.06386939]])

<span style="color:red;">**Question 7-5:** What is the shape of the computed variances?</span>

____________

# Multinomial Naive Bayes

Our second dataset (spam/not spam), our assumption is our data follows a multinomial distribution.

**Dataset:**

Our goal with this dataset is to classify a sentence as either **spam** or **not spam** (ham). 

You can check out `the spam/ham.csv` for examples of spam and not spam messages. Check the file and see its body contents.

(This section is a slight modification from <a src="http://www.ritchieng.com/machine-learning-multinomial-naive-bayes-vectorization/">Ritchie Ng's notebook</a>)

## Sample data

Before we go and train with the spam/ham dataset, we have to convert the `content` column into numbers we can crunch. In our case, our features will be the frequency of words in the data instance.

**Example:**


|                                                  | Never | gonna | give | you | up | let | down | make | cry | say | goodbye |
|--------------------------------------------------|-------|-------|------|-----|----|-----|------|------|-----|-----|---------|
|                          Never gonna give you up |   1   |   1   |   1  |  1  |  1 |  0  |   0  |   0  |  0  |  0  |    0    |
| Never gonna give you up Never gonna let you down |   2   |   2   |   1  |  2  |  1 |  1  |   1  |   0  |  0  |  0  |    0    |
|                         Never gonna make you cry |   1   |   1   |   0  |  1  |  0 |  0  |   0  |   1  |  1  |  0  |    0    |
|                          Never gonna say goodbye |   1   |   1   |   0  |  0  |  0 |  0  |   0  |   0  |  0  |  1  |    1    |

<div style="text-align: right"><sub>Reference: Never Gonna Give You Up by Rick Astley</sub></div>

In [32]:
data = ["Never gonna give you up",
        "Never gonna give you up Never gonna let you down",
        "Never gonna make you cry",
        "Never gonna say goodbye"]

First, let's convert our words all to lower case. This is a common practice.

In [33]:
for i in range(len(data)):
    data[i] = data[i].lower()
    
data

['never gonna give you up',
 'never gonna give you up never gonna let you down',
 'never gonna make you cry',
 'never gonna say goodbye']

Now, we'll count for the frequency of each word of each sentence.

We will use [CountVectorizer](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html) to convert the text into a matrix of word/token counts

In [35]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()

This following code will get the words in our dataset

In [36]:
vectorizer.fit(data)

CountVectorizer()

Let's see our words/tokens/features:

In [37]:
word_features = vectorizer.get_feature_names_out()
word_features

array(['cry', 'down', 'give', 'gonna', 'goodbye', 'let', 'make', 'never',
       'say', 'up', 'you'], dtype=object)

The following code computes for the counts of each word for each of our data sentences. 

It outputs a **sparse count matrix**. 

__Note:__ The sparse refers to the matrix having mostly 0 values for the columns (see table above). If we store this as a normal matrix, it will take up a lot of space. To save space, the following data is stored in this fashion:
> `(<sentence>, <word>)         <count>`

All combinations where the count is 0 will be ignored

In [38]:
count_sparse_matrix = vectorizer.fit_transform(data)
print(count_sparse_matrix)

<Compressed Sparse Row sparse matrix of dtype 'int64'
	with 21 stored elements and shape (4, 11)>
  Coords	Values
  (0, 7)	1
  (0, 3)	1
  (0, 2)	1
  (0, 10)	1
  (0, 9)	1
  (1, 7)	2
  (1, 3)	2
  (1, 2)	1
  (1, 10)	2
  (1, 9)	1
  (1, 5)	1
  (1, 1)	1
  (2, 7)	1
  (2, 3)	1
  (2, 10)	1
  (2, 6)	1
  (2, 0)	1
  (3, 7)	1
  (3, 3)	1
  (3, 8)	1
  (3, 4)	1


It may seem a lot of work to save little space, but as your data grows this will save a ton of memory.

In [39]:
n_sentences = count_sparse_matrix.shape[0]
n_word_features = count_sparse_matrix.shape[1]

# header
for i in range(n_word_features):
    print(word_features[i], end ="\t")
print("sentence", end="\n")
    
for i in range(n_sentences):
    for j in range(n_word_features):
        print(count_sparse_matrix[i, j], end="\t")
    print(data[i], end="\n")

cry	down	give	gonna	goodbye	let	make	never	say	up	you	sentence
0	0	1	1	0	0	0	1	0	1	1	never gonna give you up
0	1	1	2	0	1	0	2	0	1	2	never gonna give you up never gonna let you down
1	0	0	1	0	0	1	1	0	0	1	never gonna make you cry
0	0	0	1	1	0	0	1	1	0	0	never gonna say goodbye


From the [scikit-learn documentation](http://scikit-learn.org/stable/modules/feature_extraction.html#text-feature-extraction):

> In this scheme, features and samples are defined as follows:

> - Each individual token occurrence frequency (normalized or not) is treated as a **feature**.
> - The vector of all the token frequencies for a given document/sentence is considered a multivariate **sample**.

> A **corpus of documents** can thus be represented by a matrix with **one row per document/sentence** and **one column per token** (e.g. word) occurring in the corpus.

> We call **vectorization** the general process of turning a collection of text documents into numerical feature vectors. This specific strategy (tokenization, counting and normalization) is called the **Bag of Words** or "Bag of n-grams" representation. Documents are described by word occurrences while completely ignoring the relative position information of the words in the document.

For our training, we will convert count_sparse_matrix into a count_dense_matrix.

In [40]:
count_dense_matrix = count_sparse_matrix.toarray()
count_dense_matrix

array([[0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1],
       [0, 1, 1, 2, 0, 1, 0, 2, 0, 1, 2],
       [1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1],
       [0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0]])

Here is our data in a pandas DataFrame

In [41]:
pd.DataFrame(count_dense_matrix, columns=word_features, index=data)

cry  down  give  gonna  \
never gonna give you up                             0     0     1      1   
never gonna give you up never gonna let you down    0     1     1      2   
never gonna make you cry                            1     0     0      1   
never gonna say goodbye                             0     0     0      1   

                                                  goodbye  let  make  never  \
never gonna give you up                                 0    0     0      1   
never gonna give you up never gonna let you down        0    1     0      2   
never gonna make you cry                                0    0     1      1   
never gonna say goodbye                                 1    0     0      1   

                                                  say  up  you  
never gonna give you up                             0   1    1  
never gonna give you up never gonna let you down    0   1    2  
never gonna make you cry                            0   0    1  
never gonna say goodbye                             1   0    0

**Summary:**

- `vect.fit(train)` **learns the vocabulary** of the training data
- `vect.transform(train)` uses the **fitted vocabulary** to build a document-term matrix from the training data
- `vect.transform(test)` uses the **fitted vocabulary** to build a document-term matrix from the testing data (and **ignores tokens** it hasn't seen before)

# On with the spam/ham dataset
## Preprocessing

Load the text data from the csv file

In [53]:
spamham = pd.read_csv("spam_ham.csv")
spamham.dropna(inplace=True)
spamham.head(10)

type      location                                               body
0  spam  data/000/001  LUXURY WATCHES - BUY YOUR OWN ROLEX FOR ONLY $...
1  spam  data/000/002  Academic Qualifications available from prestig...
2   ham  data/000/003  Greetings all. This is to verify your subscrip...
3  spam  data/000/004  try chauncey may conferred the luscious not co...
4   ham  data/000/005  It's quiet. Too quiet. Well, how about a straw...
5   ham  data/000/006  It's working here. I have departed almost tota...
6  spam  data/000/008  The OIL sector is going crazy. This is our wee...
7  spam  data/000/009  Little magic. Perfect weekends.http://othxu.rz...
8   ham  data/000/010  Greetings all. This is a mass acknowledgement ...
9  spam  data/000/011  Hi, L C P A X V V e I r m a A I v A o b n L A ...

In [50]:
from sklearn import preprocessing

Before we proceed to vectorizing, let's change our label type from "spam" and "ham" to numerical values.

Use `sklearn`'s `LabelEncoder` to to encode the spamham dataset's labels (`type` in the csv)

In [56]:
# write code here
label_enc = preprocessing.LabelEncoder()
spamham["type"] = label_enc.fit_transform(spamham["type"])

Then, get the mapping so we know what the `0`s and `1`s mean later in the notebook

In [57]:
mapping = dict(zip(label_enc.classes_, label_enc.transform(label_enc.classes_)))
print("Mapping:", mapping)

Mapping: {'ham': np.int64(0), 'spam': np.int64(1)}


Now, call `LabelEncoder` to label encode the `type` column

In [59]:
# write code here
spamham.head(10)

type      location                                               body
0     1  data/000/001  LUXURY WATCHES - BUY YOUR OWN ROLEX FOR ONLY $...
1     1  data/000/002  Academic Qualifications available from prestig...
2     0  data/000/003  Greetings all. This is to verify your subscrip...
3     1  data/000/004  try chauncey may conferred the luscious not co...
4     0  data/000/005  It's quiet. Too quiet. Well, how about a straw...
5     0  data/000/006  It's working here. I have departed almost tota...
6     1  data/000/008  The OIL sector is going crazy. This is our wee...
7     1  data/000/009  Little magic. Perfect weekends.http://othxu.rz...
8     0  data/000/010  Greetings all. This is a mass acknowledgement ...
9     1  data/000/011  Hi, L C P A X V V e I r m a A I v A o b n L A ...

**Sanity Check:** The type column should now be in 1's and 0's. Make sure that they are still properly labelled.

Now, we will separate our features `X` from our labels `y`. Disregard the `location` column (it points to the text file where the text `body` came from)

In [61]:
# write code here
X = spamham["body"]
y = spamham["type"]

print("X shape : ", X.shape)
print("y shape : ", y.shape)

X shape :  (30974,)
y shape :  (30974,)


**Sanity Check:**
You should see the following:
```
X shape :  (30974,)
y shape :  (30974,)
```

In [62]:
print(X[0:5])
print(y[0:5])

0    LUXURY WATCHES - BUY YOUR OWN ROLEX FOR ONLY $...
1    Academic Qualifications available from prestig...
2    Greetings all. This is to verify your subscrip...
3    try chauncey may conferred the luscious not co...
4    It's quiet. Too quiet. Well, how about a straw...
Name: body, dtype: object
0    1
1    1
2    0
3    1
4    0
Name: type, dtype: int64


**Sanity Check:**
You should see the following:
```
0    LUXURY WATCHES - BUY YOUR OWN ROLEX FOR ONLY $...
1    Academic Qualifications available from prestig...
2    Greetings all. This is to verify your subscrip...
3    try chauncey may conferred the luscious not co...
4    It's quiet. Too quiet. Well, how about a straw...
Name: body, dtype: object
0    1
1    1
2    0
3    1
4    0
Name: type, dtype: int64
```

Split the dataset into train and test data sets. Set the test size to 30%, and `random_state` to 42. Make sure we also stratify based on the type (spam/ham).

In [64]:
# write code here
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42, stratify=y)

In [65]:
print("Train class distribution:\n", y_train.value_counts(normalize=True))

Train class distribution:
 type
1    0.622481
0    0.377519
Name: proportion, dtype: float64


In [66]:
print("\nTest class distribution:\n", y_test.value_counts(normalize=True))


Test class distribution:
 type
1    0.622404
0    0.377596
Name: proportion, dtype: float64


You should see that the distribution of classes in the train and test sets are maintained (1.648:1)

### Vectorization

Let's process the data as we did in the section before. Note that we will get a new dictionary based on the training dataa (we won't use the *Never gonna give you up* dataset anymore).

In [72]:
from sklearn.feature_extraction.text import CountVectorizer

count_vect = CountVectorizer()

Get the words from the training set (we should train without knowing the words from the test set)

In [73]:
X_train_counts = vectorizer.fit_transform(X_train)
X_test_counts = vectorizer.transform(X_test)
word_features = vectorizer.get_feature_names_out()

**Sanity Check:** This is a large dataset, it may take a few seconds.

And, then get the frequency of each word in  each sentence

In [74]:
X_train_count_sparse_matrix = count_vect.fit_transform(X_train)

__Note:__ A shorthand of these two lines is `count_vect.fit_transform()`

In [75]:
X_train_count_sparse_matrix.shape

(21681, 147622)

**Sanity Check:** The shape should be around (21681, 147622)

Let's check out the fitted vocabulary

In [76]:
vocab = count_vect.vocabulary_
print(vocab)

{'勝手にメール送ってしまいまして申し訳有りませんが': 145445, '期間人数限定の企画を貴方に紹介したいので': 146395, 'ご理解を頂きたく思います': 143840, '当サイトは真剣な出会空間を運営しおかげさまで好評を得ております': 145965, 'ハートもボディーもオープン的な女性を広告誌等の募集をかけ': 144669, '約１０００名の女性会員が集まりましたので': 146987, '男性会員不足の為': 146707, '貴方を含めて８００名の特選男性に特別企画を紹介致します': 147265, '企画詳細': 145120, 'http': 66788, 'www': 140803, 'otxv': 95351, 'com': 38497, 'num': 91326, '508〇リアルタイム紹介出会えるまでの時間を短縮する為': 10785, '当サイトは': 145960, '特別無料紹介機能': 146626, 'を利用しており': 144410, 'リアルタイムでやり取りできる女性のみを紹介していくという機能です': 144785, 'メールしても返事は返ってきません': 144756, 'とかの心配は一切ありません': 144137, 'この機能により': 143758, '最短数分で気に入った女性と出会えます': 146354, '直アド': 146794, '写メ配信只今男性会員不足なので': 145340, '女性会員から逆指名されたケースが非常に急増しており': 145691, '当サイトは貴方を指名した女性会員のメッセージだけではなく': 145966, '直アドと写メ': 146795, '公開中': 145321, 'も無料で貴方に配信致します': 144364, '面倒な検索は一切不要です': 147487, 'この企画逃したら勿体無い': 143739, '508': 10781, '選男性専用': 147405, '迷惑だったら': 147339, 'please': 99794, 'inform': 69755, 'the': 124004, 'following': 56890, 'address': 18559, 'if': 67900, 'this': 127035, 'mail': 81755,

**Sanity Check:** This will really get funny characters. Try seeing the 45,000th words onward to see more "normal" words

In [84]:
sorted_vocab = sorted(vocab.items(), key=lambda x: x[1])
print(sorted_vocab[45000:45050])


[('demütige', 45000), ('den', 45001), ('denardo', 45002), ('denature', 45003), ('dendai', 45004), ('dendritic', 45005), ('denebola', 45006), ('denen', 45007), ('denesh', 45008), ('deneyse', 45009), ('deng', 45010), ('deniability', 45011), ('deniable', 45012), ('denial', 45013), ('denialattacks', 45014), ('denialkurtz', 45015), ('denicola', 45016), ('denied', 45017), ('denied2006', 45018), ('deniedcause', 45019), ('denies', 45020), ('denikzaslotce', 45021), ('denim', 45022), ('denis', 45023), ('denise', 45024), ('denizen', 45025), ('denk', 45026), ('denke', 45027), ('denkt', 45028), ('denmark', 45029), ('denmarkduring', 45030), ('denn', 45031), ('denning', 45032), ('dennis', 45033), ('dennisgraduate', 45034), ('dennison', 45035), ('dennispaulson', 45036), ('dennist2', 45037), ('dennistdennist2', 45038), ('dennisyou', 45039), ('dennoch', 45040), ('deno', 45041), ('denoeux', 45042), ('denograduate', 45043), ('denominate', 45044), ('denominator', 45045), ('denotational', 45046), ('denotati

Now, we also have to transform our test data to our fitted vocab. 


**Note:** We should not fit the test data's vocabulary. We're going to use the word features we culled from the training dataset.

In [85]:
# write code here
X_test_count_sparse_matrix = count_vect.transform(X_test)

X_test_count_sparse_matrix.shape

(9293, 147622)

**Sanity Check:**
The number of features (dimensions, not instances) of the train and test should match

**Now we have two transformed sparse matrices:**
- X_train_count_sparse_matrix
- X_test_count_sparse_matrix



## Modelling

Now that we've got preprocessing done, we can focus on building the model. Here, we will use sklearn's `MultinomialNB` because our assumption is that our data follows a multinomial distribution.

In [86]:
from sklearn.naive_bayes import MultinomialNB
# MultinomialNB?

Initialize your `MultinomialNB` model

In [87]:
spam_nb = MultinomialNB()

Fit it to out training data (`X_train_count_sparse_matrix`)

In [88]:
spam_nb.fit(X_train_count_sparse_matrix, y_train)

MultinomialNB()

And, get our training predictions

In [89]:
predictions = spam_nb.predict(X_train_count_sparse_matrix)
predictions

array([1, 1, 0, ..., 1, 1, 1])

Let's see how well our model worked on our training data

In [90]:
print("Spam/ham training accuracy: ", compute_accuracy(predictions, y_train), "%")

Spam/ham training accuracy:  0.9904985932383192 %


Then, let's see how it does on our test set

In [91]:
predictions = spam_nb.predict(X_test_count_sparse_matrix)
predictions

array([1, 0, 1, ..., 1, 0, 1])

In [92]:
print("Spam/ham test accuracy: ", compute_accuracy(predictions, y_test), "%")

Spam/ham test accuracy:  0.9821370924351662 %


**Sanity Check:** You should get around a 98% accuracy

We should also be able to call `classification_report` to see how well our model performed with different metrics

In [93]:
from sklearn.metrics import classification_report
# classification_report?

Print the test classification report of our model. Set the `target_names` to `mapping.keys()` so we can see what `0` and `1` refers to.

In [94]:
# write code here
print(classification_report(y_test, predictions, target_names=mapping.keys()))


              precision    recall  f1-score   support

         ham       0.96      1.00      0.98      3509
        spam       1.00      0.97      0.99      5784

    accuracy                           0.98      9293
   macro avg       0.98      0.99      0.98      9293
weighted avg       0.98      0.98      0.98      9293



**Sanity check:** You should get the following results
```
              precision    recall  f1-score   support

         ham     0.9573    0.9972    0.9768      3509
        spam     0.9982    0.9730    0.9855      5784

    accuracy                         0.9821      9293
   macro avg     0.9778    0.9851    0.9811      9293
weighted avg     0.9828    0.9821    0.9822      9293
```

<span style="color:red;">**Question 7-6:** Among the classes (`ham` or `spam`), which is more likely to get labelled its class?</span>

## Testing our model with our own input

In [100]:
input_test = input("Enter text to check if spam or ham : ")
while input_test.lower() != "q":

    input_test_matrix = count_vect.transform([input_test])

    results = spam_nb.predict(input_test_matrix)
    results_label = ["HAM", "SPAM"]
    print("Text : " + input_test + " is " + results_label[results[0]])

    input_test = input("Enter text to check if spam or ham : ")

Text : congratulations you won a free iphone! is SPAM
Text : hey, are we meeting later? is HAM


## Checking the learned parameters
Let's see the parameters the `MultinomialNB` model learned.

Get the token counts the model computed

In [101]:
token_counts = spam_nb.feature_count_
token_counts.shape

(2, 147622)

**Sanity check:** You should get a `(2, 147622)` matrix

<span style="color:green;">Question: Why did we get a `(2, 147622)` matrix for the token counts?</span>

To get the token counts of `spam` or `ham`, we can use our `mapping`.

In [102]:
spam_token_counts = token_counts[mapping["spam"]]
ham_token_counts = token_counts[mapping["ham"]]

We can sort the token counts to see the word that occurs the less/most for that class

In [ ]:
sorted_spam_counts = np.sort(spam_token_counts)
sorted_ham_counts = np.sort(ham_token_counts)

While `np.sort` returns the actual counts, `np.argsort` returns the sorted indices

In [118]:
spam_idx = np.argsort(spam_token_counts)
ham_idx = np.argsort(ham_token_counts)

print(spam_idx)


[    18     19     20 ... 128265  21596 124004]


**Sanity check:** You should see the following:

`array([ 73810, 119826, 119827, ..., 128265,  21596, 124004])`

The two sorts show that the `73,810th` word occurred `0` times, while the `124,004th` occurred `53,076` times in the spam sentences. Note that these are raw counts that are skewed because there are significantly more spam sentences. **The model normalizes the counts relative to the class.**

To get the the `ith` word/token, we can use `count_vect.get_feature_names()`

In [119]:
feature_names = count_vect.get_feature_names_out()
most_frequent_spam_word_index = sorted_spam_indices[-1]
most_frequent_spam_word = feature_names[most_frequent_spam_word_index]
print(most_frequent_spam_word)

the


<span style="color:red;">Question 7-7: What word occurred the most in the spam sentences? Write the word in small letters only.</span>

The following code lists the top occurring words per class:

In [121]:
top = 50


print("spam \t ham")
print("------------")

for i in range(top):
    print(count_vect.get_feature_names_out()[ham_idx[i]], "\t", count_vect.get_feature_names_out()[spam_idx[i]])

spam 	 ham
------------
ｓｅｘを求めている 	 0000006hz
ｓ子様秘密が条件で 	 000000eb
ｓクラス専門店 	 000001
ｐｃから簡単プロフィール作成 	 00000100
ｐｃ 	 00000100shaftencoder
ｏｌ 	 00000111
ｎ藤 	 000001bdaaa0
ｍａｉｌでのサポートは２４時間対応です 	 000001bdb744
ｍ字開脚オナニーを机の下から盗撮 	 000001bdc5a5
baden 	 000001bdd411
bade 	 000001bdd98c
bacteriaphagocyte 	 000001bdda70
bacokcyzi 	 000001bde0a0
bagger 	 000001bed6b7
baggages 	 recherches
bagentintroduced 	 recherhe
ｈのお相手しただけで 	 按鈕2
ｈにそんなに興味なかったのと少し怖いのもあるため 	 滑鼠
ｈな欲望や願望を胸に秘め 	 鍵盤
ｈなこと大好きな人ばかり 	 0000
ｇｗ特典あり 	 00000000
ｇｗはこういう女の子と過ごしたいっす 	 000000000
ｇｗです 	 00000000000000
ｇｗこそ出会いのチャンスhttp 	 0000000016
ｇｒａｎｄｅｅの理念やシステムのご紹介 	 00000000x
ｇｒａｎｄｅｅ 	 00000001
ｇｏｏｄ 	 00000001content
000001c64310 	 00000001irdecode
000001c642d0 	 00000004
ｔ谷 	 00000010
ｗ６２ 	 00000010pwm
ｓｅｘを求めているのです 	 00000049
２６日 	 0000100c
２４歳 	 00001010
２４才 	 00001014
２２歳 	 00001018
２２日 	 0000101c
２１歳 	 00001023
ｇａｌ誌多数掲載 	 00001026
ｆカップ 	 00001028
ｂ９６ 	 0000102c
ａ美様現在未亡人でいらっしゃいます 	 00001030
ａ型 	 00001034
６名のみとなります 	 00001038
５００円分のポイントが完全無料で自

<s>With the code above, you can now reword your scam so that it can bypass a common spam filter.</s>

The model does not depend on raw counts but instead uses the log probability. Get the model's log probabilities.

In [122]:
# write code here
log_prob = spam_nb.feature_log_prob_

Let's sort the `feature_log_prob_` similar to the way we sorted the token counts

In [123]:
spam_log_prob = log_prob[mapping["spam"]]
ham_log_prob = log_prob[mapping["ham"]]

In [126]:
spam_log_idx = np.argsort(spam_log_prob)
ham_log_idx = np.argsort(ham_log_prob)

spam_log_idx = spam_log_idx[::-1]
ham_log_idx = ham_log_idx[::-1]

feature_names = count_vect.get_feature_names_out()

top = 50
print("ham \t spam")
print("------------")

for i in range(top):
    print(feature_names[ham_log_idx[i]], "\t", feature_names[spam_log_idx[i]])

ham 	 spam
------------
the 	 the
to 	 and
of 	 to
and 	 of
in 	 in
is 	 you
for 	 is
it 	 font
that 	 http
you 	 for
on 	 this
this 	 padding
with 	 our
be 	 it
from 	 your
have 	 with
are 	 we
at 	 com
as 	 0px
not 	 that
if 	 price
or 	 border
by 	 product_table
can 	 on
but 	 are
edu 	 from
will 	 color
an 	 top
my 	 95
we 	 size
your 	 as
all 	 weight
one 	 at
would 	 be
there 	 will
was 	 info
any 	 by
so 	 my
http 	 or
use 	 all
has 	 not
do 	 none
com 	 left
they 	 have
what 	 right
which 	 more
about 	 company
some 	 no
www 	 www
list 	 out


We can see that the order is maintained, and the `124,004th` word is still the most occurring word in the `spam` sentences.

We can also see the class count and computed priors for each class

In [127]:
print("Class Count:", spam_nb.class_count_)

Class Count: [ 8185. 13496.]


In [128]:
print("Class Log Prior:", spam_nb.class_log_prior_)

Class Log Prior: [-0.97413309 -0.47404296]


Note that the priors are computed based on the count of each class (spam or not spam) in the dataset. The log probability is computed.

# Tuning our Naive Bayes model

In this section we will reuse our spam/ham dataset. We will resplit our dataset in the following manner:
1. Allot 20% of the original dataset as our hold-out test set.
1. Allot 25% of our remaining data as our validation data set. The remaining 80% will serve as our training data.

We will use `sklearn`'s `ParameterGrid` to tune our hyperparameters

Let's separate our test set. Set the test set to `20%`, stratify based on the target class, and set the `random_state` to 42.

In [129]:
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=.2, random_state=42, stratify=y)

print("X_test shape : ", X_test.shape)
print("y_test shape : ", y_test.shape)

X_test shape :  (6195,)
y_test shape :  (6195,)


We will the same thing to separate our validation set. Set the validation set size to `25%`, stratify based on the target class, and set the `random_state` to 42. 

__Don't forget that we are now splitting `X_train_val` and `y_train_val`.__ There should be no data leakage.

In [132]:
X_train, X_validation, y_train, y_validation = train_test_split(X_train_val, y_train_val, test_size=.25, random_state=42, stratify=y_train_val)

print("X_train shape : ", X_train.shape)
print("y_train shape : ", y_train.shape)
print("X_validation shape : ", X_validation.shape)
print("y_validation shape : ", y_validation.shape)

X_train shape :  (18584,)
y_train shape :  (18584,)
X_validation shape :  (6195,)
y_validation shape :  (6195,)


## Vectorization

Now that we have our data sets prepared, we can now start computing for the token counts. Remember that we will have to refit our vectorizer to our new train data.

Initialize a `CountVectorizer`. Set it to remove English stop words. This should remove common words like `the`, `of`, `and` that likely will not anything meaningful to distinguish the two classes apart.

In [133]:
# write code here
count_vect = CountVectorizer(stop_words="english")

Get the count matrix for the training and validation set. 

In [134]:
X_train_count_sparse_matrix = count_vect.fit_transform(X_train)
X_validation_count_sparse_matrix = count_vect.transform(X_validation)

In [135]:
print("X_train_count_sparse_matrix shape: ", X_train_count_sparse_matrix.shape)
print("X_validation_count_sparse_matrix shape: ", X_validation_count_sparse_matrix.shape)

X_train_count_sparse_matrix shape:  (18584, 146302)
X_validation_count_sparse_matrix shape:  (6195, 146302)


**Sanity check:** You should see the following values:
```
X_train_count_sparse_matrix shape:  (18584, 146302)
X_validation_count_sparse_matrix shape:  (6195, 146302)
```

<span style="color:red;">**Question 7-8:** Why should we not get the count vectorizer to fit on the validation data set instead?</span>

## GridSearch with `ParameterGrid`
In this section, we will use `ParameterGrid` to get the combinations of hyperparameters we will try on our model.

In [136]:
from sklearn.model_selection import ParameterGrid

Set our base classifier for the spam/ham classifier. Don't train yet

In [137]:
# write code here
spam_nb = MultinomialNB()

For this model, we can tweak the `alpha` (our smoothing operator) and whether or not we want to compute for the prior (`fit_prior`). You can read more about this in the docs.

In [139]:
...

Ellipsis

For the following section, we will define our hyperparameters. For now, set the following hyperparameter choices:

__Hyperparameters__:
- alpha could be 1, 3, 5, 10, 15, 20, 50
- fit_prior could be true or false

In [138]:
# write code here
hyperparameters = [{
    'alpha': [1, 3, 5, 10, 15, 20, 50],
    'fit_prior': [True, False]
}]

If we call `ParameterGrid`, it should list the following:

In [140]:
list(ParameterGrid(hyperparameters))

[{'alpha': 1, 'fit_prior': True},
 {'alpha': 1, 'fit_prior': False},
 {'alpha': 3, 'fit_prior': True},
 {'alpha': 3, 'fit_prior': False},
 {'alpha': 5, 'fit_prior': True},
 {'alpha': 5, 'fit_prior': False},
 {'alpha': 10, 'fit_prior': True},
 {'alpha': 10, 'fit_prior': False},
 {'alpha': 15, 'fit_prior': True},
 {'alpha': 15, 'fit_prior': False},
 {'alpha': 20, 'fit_prior': True},
 {'alpha': 20, 'fit_prior': False},
 {'alpha': 50, 'fit_prior': True},
 {'alpha': 50, 'fit_prior': False}]

For every iteration, we will:
1. Set the parameters of our base model to the current hyperparameter combination 
1. Fit our model to our training data
1. Compute for our training accuracy
1. Run predictions on our validation data
1. Compute for our training accuracy
1. Keep track of the best performing validation accuracy and its associate hyperparam combo.

In [141]:
from sklearn.metrics import accuracy_score

best_score = 0
for g in ParameterGrid(hyperparameters):
    print(g)
    
    spam_nb.set_params(**g)
    
    # write code here
    spam_nb.fit(X_train_count_sparse_matrix, y_train)
    train_predictions = spam_nb.predict(X_train_count_sparse_matrix)
    train_acc = accuracy_score(y_train, train_predictions) * 100
    
    # write code here
    val_predictions = spam_nb.predict(X_validation_count_sparse_matrix)
    val_acc = accuracy_score(y_validation, val_predictions) * 100
    
    print(f"Train acc: {train_acc}% \t Val acc: {val_acc}%", end="\n\n")
    
    if val_acc > best_score:
        best_score = val_acc
        best_grid = g

print("Best accuracy: ", best_score, "%")
print("Best grid: ", best_grid)

{'alpha': 1, 'fit_prior': True}
Train acc: 99.08523461041756% 	 Val acc: 98.4503631961259%

{'alpha': 1, 'fit_prior': False}
Train acc: 99.02066293585881% 	 Val acc: 98.36965294592413%

{'alpha': 3, 'fit_prior': True}
Train acc: 98.63861386138613% 	 Val acc: 98.15980629539952%

{'alpha': 3, 'fit_prior': False}
Train acc: 98.59018510546707% 	 Val acc: 98.07909604519774%

{'alpha': 5, 'fit_prior': True}
Train acc: 98.40723202755058% 	 Val acc: 98.01452784503631%

{'alpha': 5, 'fit_prior': False}
Train acc: 98.32651743435214% 	 Val acc: 97.88539144471348%

{'alpha': 10, 'fit_prior': True}
Train acc: 97.98213517003875% 	 Val acc: 97.67554479418887%

{'alpha': 10, 'fit_prior': False}
Train acc: 97.82070598364184% 	 Val acc: 97.38498789346247%

{'alpha': 15, 'fit_prior': True}
Train acc: 97.72384847180369% 	 Val acc: 97.49798224374496%

{'alpha': 15, 'fit_prior': False}
Train acc: 97.69156263452432% 	 Val acc: 97.30427764326069%

{'alpha': 20, 'fit_prior': True}
Train acc: 97.65389582436505%

<span style="color:red;">Question 7-9: What is the best found value for `alpha`? Round off to four decimal places.</span>

<span style="color:red;">Question 7-10: What is the best found value for `fit_prior`?</span>

## Retraining our estimator with the best hyperparameters

Now that we know the best hyperparameters, we can now make a new classifier and retrain it.

In [142]:
# write code here
spam_nb = MultinomialNB(alpha=1, fit_prior=True)

Make sure you train it with both our training and validation set. You can keep the trained `count_vect`.

In [143]:
# write code here
X_train_val_count_sparse_matrix = count_vect.transform(X_train_val)

In [144]:
# write code here
spam_nb.fit(X_train_val_count_sparse_matrix, y_train_val)

MultinomialNB(alpha=1)

## Testing phase

Run predictions on the test data set

In [145]:
X_test_val_count_sparse_matrix = count_vect.transform(X_test)
predictions = spam_nb.predict(X_test_val_count_sparse_matrix)

Compute for the test accuracy

In [146]:
test_acc = compute_accuracy(predictions, y_test)
print("Test accuracy: ", test_acc, "%")

Test accuracy:  0.9830508474576272 %


<span style="color:red;">Question 7-11: What is the final test accuracy? Round off to four decimal places.</span> 

# Summary

In this notebook, we created two kinds of Naive Bayes models: Gaussian and Multinomial. 

We also saw the models' learned parameters. For Gaussian NB models, the model learns the mean and standard deviation of each feature per class, while multinomial NB models learn the log probability of each token per class.

We also experienced creating a natural language processing (NLP) machine learning model. Unlike its deep learning counterpart, the features are more hand-crafted because we dictate what the model should look at. In this case, we specifically designed it to look at token/term frequency/count, but we could build more sophisticated versions like inverse document frequency or term frequency-inverse document frequency (TF-IDF). 

## <center>fin</center>


<!-- DO NOT MODIFY OR DELETE THIS -->

<sup>made/compiled by daniel stanley tan & courtney anne ngo 🐰 & thomas james tiam-lee</sup> <br>
<sup>for comments, corrections, suggestions, please email:</sup><sup> danieltan07@gmail.com & courtneyngo@gmail.com & thomasjamestiamlee@gmail.com</sup><br>
<sup>please cc your instructor, too</sup>
<!-- DO NOT MODIFY OR DELETE THIS -->